app/controllers/nonsipp/otherassetsheld/IncomeFromAssetController.scala

```diff
@@ -0,0 +1,120 @@
+
+import config.Constants
+import config.Refined.Max5000
+import controllers.PSRController
+import controllers.actions._
+import controllers.nonsipp.otherassetsheld.IncomeFromAssetController._
+import forms.MoneyFormProvider
+import forms.mappings.errors.MoneyFormErrors
+import models.SchemeId.Srn
+import models.{Mode, Money}
+import navigation.Navigator
+import pages.nonsipp.otherassetsheld.IncomeFromAssetPage
+import play.api.data.Form
+import play.api.i18n.MessagesApi
+import play.api.mvc.{Action, AnyContent, MessagesControllerComponents}
+import services.SaveService
+import utils.FormUtils._
+import viewmodels.DisplayMessage.{Empty, Message}
+import viewmodels.implicits._
+import viewmodels.models.MultipleQuestionsViewModel.SingleQuestion
+import viewmodels.models.{FormPageViewModel, QuestionField, SectionCompleted}
+import views.html.MoneyView
+
+import javax.inject.{Inject, Named}
+import scala.concurrent.{ExecutionContext, Future}
+
+class IncomeFromAssetController @Inject()(
+  override val messagesApi: MessagesApi,
+  @Named("non-sipp") navigator: Navigator,
+  identifyAndRequireData: IdentifyAndRequireData,
+  saveService: SaveService,
+  formProvider: MoneyFormProvider,
+  val controllerComponents: MessagesControllerComponents,
+  view: MoneyView
+)(implicit ec: ExecutionContext)
+    extends PSRController {
+
+  private def form = IncomeFromAssetController.form(formProvider)
+
+  def onPageLoad(srn: Srn, index: Max5000, mode: Mode): Action[AnyContent] = identifyAndRequireData(srn) {
+    implicit request =>
+      Ok(
+        view(
+          viewModel(srn, index, form.fromUserAnswers(IncomeFromAssetPage(srn, index)), mode)
+        )
+      )
+
+  }
+
+  def onSubmit(srn: Srn, index: Max5000, mode: Mode): Action[AnyContent] = identifyAndRequireData(srn).async {
+    implicit request =>
+      form
+        .bindFromRequest()
+        .fold(
+          formWithErrors =>
+            Future.successful(
+              BadRequest(
+                view(
+                  IncomeFromAssetController.viewModel(srn, index, formWithErrors, mode)
+                )
+              )
+            ),
+          answer => {
+            for {
+              updatedAnswers <- Future
+                .fromTry(
+                  request.userAnswers
+                    .set(IncomeFromAssetPage(srn, index), answer)
+                )
+              _ <- saveService.save(updatedAnswers)
+            } yield Redirect(navigator.nextPage(IncomeFromAssetPage(srn, index), mode, updatedAnswers))
+          }
+        )
+  }
+}
+
+object IncomeFromAssetController {
+  def form(formProvider: MoneyFormProvider): Form[Money] = formProvider(
+    MoneyFormErrors(
+      "otherAssets.incomeFromAsset.error.required",
+      "otherAssets.incomeFromAsset.error.invalid",
+      (Constants.maxMoneyValue, "otherAssets.incomeFromAsset.error.tooLarge"),
+      (Constants.zeroMoneyValue, "otherAssets.incomeFromAsset.error.tooSmall")
```

Comment:
I'm wondering if it would be better to reuse `"otherAssets.incomeFromAsset.error.invalid"` here instead of `"otherAssets.incomeFromAsset.error.tooSmall"`, as the latter doesn't match a key in `messages.en` - what do you think? An alternative would be to create the key `"otherAssets.incomeFromAsset.error.tooSmall"` with the same error message as `"otherAssets.incomeFromAsset.error.invalid"`.

I appreciate this is sort of irrelevant, as any negative input will trigger the invalid input error message anyways, so this definitely shouldn't block this ticket from being merged in, but making a reference to a key that doesn't exist in `messages.en` stuck out to me, so I thought I'd raise this question.

Reply:
I agree with reusing `otherAssets.incomeFromAsset.error.invalid`, and will correct this in the current branch I am working on.